### ISSUE TITLE
job-manager: use JSON in event context

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#2076]
The ad-hoc metadata form used in eventlogs (key=value key=value...) is limited in what it can represent, and is more painful to parse than you would think.

[RFC 18](https://github.com/flux-framework/rfc/blob/master/spec_18.adoc) doesn't specify any format for the context field of an event - it's just a free form string there.  [RFC 21](https://github.com/flux-framework/rfc/blob/master/spec_21.adoc) does, however, specify the above format.

I propose we use serialized JSON here (serialized without newlines) and update RFC 21 to require it in the job manager events that drive state transitions.  The immediate benefits are:
* The exec `release` "final" key can be represented as a JSON boolean as it is in the message payload
* Strings with spaces, equals, and newlines can be represented
* Eliminate `util_int_from_context()`, `util_note_from_context()`,and ` util_str_from_context()`

**Comment by chu11** [comment]
> I propose we use serialized JSON here (serialized without newlines)

I think I'm missing something.  The newline is part of the event, not the context.  You're only proposing serialized json for the context?  So the newline is still needed as a delimeter between events?

**Comment by grondo** [comment]
I think @garlick meant the JSON part would have to be serialized without any non-escaped newlines (for just the reason you mention).

I assume the eventlog parser itself would  then have to be aware of how newlines are escaped...

**Comment by garlick** [comment]
Right.  Json newlines are represented as \n I think so no need to take them into account when parsing eventlogs since the event delimiter is literal newline.

**Comment by grondo** [comment]
> Json newlines are represented as \n I think so no need to take them into account when parsing eventlogs since the event delimiter is literal newline.

Oh yeah duh!

**Comment by chu11** [comment]
> I think @garlick meant the JSON part would have to be serialized without any non-escaped newlines (for just the reason you mention).

Duh, I see how I misread it now.

**Comment by garlick** [comment]
I'll go ahead and submit a PR to update RFC 21 to specify JSON for the events in the primary job eventlog.

I think it's appropriate to leave RFC 18 open ended.

**Comment by chu11** [comment]
Was pondering the effect this change should have on commands like `flux job eventlog`, perhaps output should still be done `key=value` for readability?  And for `flux kvs eventlog append` it should take `key=value` pairs as input?

**Comment by grondo** [comment]
Good question! I was kind of pondering the same thing, but not really coming to any solid conclusions.

Keeping the output the same for now would be kind of slick (not that anything is really parsing the output, but it *is* slightly more readable as you say).  Maybe with an option to keep the output in JSON (an extensible option like `--context-format=json|text|...`?) The idea that the context is in JSON opens some interesting possibilities (i.e. you could eventually match not only an entry in the log but some component of its context too), but I can't think of any actual use cases for that right now.

**Comment by chu11** [comment]
Makes sense, to have a `--context-format=...` for a `flux kvs job`. 

Now that I think about it, since this is specifically for jobs, it probably wouldn't be appropriate for changes to `flux kvs eventlog append`.  Users of that would have to pre-serialize any data going into a job eventlog.  Similarly for `flux kvs eventlog get` it probably should output the raw output.

**Comment by garlick** [comment]
Agreed, at least I was proposing that we leave RFC 18 agnostic about any context formatting, so if it stays that way, the generic `flux kvs eventlog` subcommand should also be agnostic IMHO.

**Comment by garlick** [comment]
Going the other way, would it make sense to have an option to emit the event as a JSON object, e.g.
```json
{"timestamp":42.33333,"name":"submit","context":{"userid":5588,"priority":16,"flags":0}}
```
Then you could easily pick apart the result in tests with `jq`...

**Comment by chu11** [comment]
closed via #2085
